chain/chain/src/chain.rs

```diff
@@ -2765,13 +2782,47 @@ impl<'a> ChainUpdate<'a> {
         })
     }
 
+    /// Applies chunks in separate threads, awaits finishing and processes results
     fn apply_chunks(
         &mut self,
         me: &Option<AccountId>,
         block: &Block,
         prev_block: &Block,
         mode: ApplyChunksMode,
     ) -> Result<(), Error> {
+        let work = self.apply_chunks_preprocessing(me, block, prev_block, mode)?;
+        self.do_apply_chunks(block, prev_block, work)
+    }
+
+    /// Applies chunks and processes results
+    fn do_apply_chunks(
+        &mut self,
+        block: &Block,
+        prev_block: &Block,
+        work: Vec<Box<dyn FnOnce() -> Result<ApplyChunkResult, Error> + Send + 'static>>,
+    ) -> Result<(), Error> {
+        work.into_par_iter().map(|task| task()).collect::<Vec<_>>().into_iter().try_for_each(
+            |result| -> Result<(), Error> {
+                self.process_apply_chunk_result(
+                    result?,
+                    block.hash().clone(),
+                    prev_block.hash().clone(),
+                )
+            },
+        )
+    }
+
+    /// Creates jobs that would apply chunks
+    fn apply_chunks_preprocessing(
+        &mut self,
+        me: &Option<AccountId>,
+        block: &Block,
+        prev_block: &Block,
+        mode: ApplyChunksMode,
+    ) -> Result<Vec<Box<dyn FnOnce() -> Result<ApplyChunkResult, Error> + Send + 'static>>, Error>
```

Comment:
Summarising zulip discussion: it's a bit unfortunate that we are returning a closure here, instead of some specific things. The reason for that is that what we want to do here is just pack arguments for `RuntimeAdapter::apply_transactions` into a single struct, and there are a lot of arguments there ([seen worse](https://github.com/JetBrains/intellij-community/blob/9c92f6e34aa959cbd8b3452a28898a7c63052915/platform/platform-impl/src/com/intellij/ui/BalloonImpl.java#L326-L354) though :) ). 

One potential solution here is to change `fn apply_transactions` to 

```
    fn apply_transactions(&self, transacions_to_apply: TransactionsToApply)
```

but:

* it's unclear if its worth it
* this is out of scope for this PR.